optimism/src/mips/mod.rs

```diff
@@ -15,9 +15,176 @@
 //! [crate::mips::interpreter], and the evaluations will be kept in the
 //! structure ProofInputs.
 
+use std::collections::HashMap;
+
+use ark_ff::Field;
+use kimchi_msm::witness::Witness;
+use strum::EnumCount;
+
+use crate::{
+    mips::{
+        column::MIPS_COLUMNS,
+        constraints::Env,
+        interpreter::Instruction::{self, *},
+    },
+    Circuit, CircuitTrait,
+};
+
+use self::interpreter::{
+    interpret_instruction,
+    ITypeInstruction::{self, *},
+    JTypeInstruction::{self, *},
+    RTypeInstruction::{self, *},
+};
+
 pub mod column;
 pub mod constraints;
 pub mod folding;
 pub mod interpreter;
 pub mod registers;
 pub mod witness;
+
+#[allow(dead_code)]
+/// The Keccak circuit
+pub type MIPSCircuit<F> = Circuit<MIPS_COLUMNS, Instruction, F>;
+
+pub const INSTRUCTIONS: [Instruction;
```

Comment:
Was this created from scratch or moved from somewhere?...

Reply:
From scratch by enumerating all the variants for the Instruction enum. If you know of a method that automatically iterates over them all please let me know! This is awful 🥲

Reply:
Nice, didn't know about this mechanism, looks quite elegant!